File: 981.time-based-key-value-store.py
#
# @lc app=leetcode id=981 lang=python3
#
# [981] Time Based Key-Value Store
#

# @lc code=start
class TimeMap:

    def __init__(self):
        from collections import defaultdict
        self.kvstore = defaultdict(lambda: [])

    def set(self, key: str, value: str, timestamp: int) -> None:
        target = self.kvstore[key]
        l, r, idx = 0, len(target)-1, len(target)
        while l <= r:
            m = l + (r - l) // 2
            if target[m][0] >= timestamp:
                idx = m
                r = m - 1
            else:
                l = m + 1
        if idx < len(target):
            if target[idx][0] == timestamp:
                target[idx][1] = value
            else:
                target.insert(idx, [timestamp, value])
        else:
            target.append([timestamp, value])

    def get(self, key: str, timestamp: int) -> str:
        target = self.kvstore[key]
        l, r, idx = 0, len(target)-1, len(target)
        while l <= r:
            m = l + (r - l) // 2
            if target[m][0] <= timestamp:
                idx = m
                l = m + 1
            else:
                r = m - 1
        if idx < len(target):
            return target[idx][1]
        else:
            return ""


# Your TimeMap object will be instantiated and called as such:
# obj = TimeMap()
# obj.set(key,value,timestamp)
# param_2 = obj.get(key,timestamp)
# @lc code=end

File: 981.time-based-key-value-store.java
import java.util.HashMap;

/*
 * @lc app=leetcode id=981 lang=java
 *
 * [981] Time Based Key-Value Store
 */

// @lc code=start

class Data {
    public String value;
    public int timestamp;

    public Data(String value, int timestamp) {
        this.value = value;
        this.timestamp = timestamp;
    }
}

class TimeMap {
    private HashMap<String, ArrayList<Data>> tm;

    public TimeMap() {
        this.tm = new HashMap<String, ArrayList<Data>>();
    }

    public void set(String key, String value, int timestamp) {
        if (!this.tm.containsKey(key)) {
            this.tm.put(key, new ArrayList<Data>());
        }
        this.tm.get(key).add(new Data(value, timestamp));
    }

    public String get(String key, int timestamp) {
        if (!this.tm.containsKey(key)) {
            return "";
        }
        return this.binarySearch(this.tm.get(key), timestamp);
    }

    private String binarySearch(ArrayList<Data> arr, int time) {
        int lp = 0;
        int rp = arr.size() - 1;
        int rtn = -1;
        while (lp <= rp) {
            int mp = lp + (rp - lp) / 2;
            if (arr.get(mp).timestamp <= time) {
                rtn = mp;
                lp = mp + 1;
            } else {
                rp = mp - 1;
            }
        }
        if (rtn == -1) {
            return "";
        }
        return arr.get(rtn).value;
    }
}

/**
 * Your TimeMap object will be instantiated and called as such:
 * TimeMap obj = new TimeMap();
 * obj.set(key,value,timestamp);
 * String param_2 = obj.get(key,timestamp);
 */
// @lc code=end
